content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+- Currently, programmers cannot declare smaller regions within the function scope. Currently, each function has a region and each pointer argument has it's own region. The next step would be to allow the programmer to create their own scoped regions. This was set asider during the project proposal process but would be interesting further work.
+- As a consequence of the first two, a programmer cannot currently allocate memory in a loop. The ideal way to do this would to have the loop body be one region so that at the end of each loop iteration any memory allocated during that iteration is freed. I attempted to infer small regions just for this use case and found it to be both difficult infer and confusing to use. It can be very unclear where the analysis has chosen to start and end a region which makes for a pour experience when the programmer can't tell when a memory location goes out of scope.
+- Every Owning pointer must dominate the exit block(s) of it's region. In this way, you cannot conditionally allocate or free memory in a Bril Program. A given allocation of memory must exist in the same variable no matter which block it exits the region from. A borrowed pointer can be freely used as long as it is invalidated once the region scope ends.
+- The programmer can specify when two arguments to a function should have the same region but they are unable to declare one region as nested within another. This would allow finer grain control over what input regions are acceptable for a function can instead the coarseness of, "Either they are the same region or they aren't".
+- A pointer returned from a ```Call``` value instruction is assumed to own the memory it is returning. This is to make a clear distinction of who's job it is to free memory so that different functions don't double free the same memory at the end of their regions. This is more of an implementation limitation to make inference easier than a technical limitation.
+
+## Implementation of the analysis
+
+The analysis is broken up into 5 stages. First the Bril program is broken up into a control flow graph of basic blocks. This a straight forward phase which involves inserting extra labels which will show up in the resulting code. An inference pass is then run on each function to add in any region annotations that may not have been included. The function is given a region name if the programmer does not specify one. This is also when the program identifies the owner pointers from the borrowed pointers. From this point, regardless of user input, the control flow graph for each function is now fully annotated and consistent. Because this inference needs to be conservative, each argument that a program takes in is assumed to be a part of its own region. This is to avoid some kind of leakage bug where a pointer of short lived memory is stored in a long living array and it ends up leaving a gap after it is free.
```

Comment:
The middle of this process could use more detail:

> An inference pass is then run on each function to add in any region annotations that may not have been included. [...] This is also when the program identifies the owner pointers from the borrowed pointers.

This is kind of the "meat" of what you have done in this project, so I think it deserves some more expansion. The other steps in this chain, such as constructing a CFG, are generic and therefore not terribly germane. But we want to know what these analyses actually do! Is it a data flow analysis? If so, what are the rules? If not, how did you traverse the CFG? Are there any pitfalls that could trip up the analysis if you weren't careful?